wast/scanner.go

```diff
@@ -0,0 +1,565 @@
+// Copyright 2017 The go-interpreter Authors.  All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package wast
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"strconv"
+	"strings"
+	"unicode"
+	"unicode/utf8"
+
+	"github.com/go-interpreter/wagon/wasm"
+)
+
+type Scanner struct {
+	file  string
+	inBuf *bytes.Buffer
+
+	ch    rune
+	eof   bool
+	token *Token
+
+	offset int
+
+	Line   int
+	Column int
+
+	Errors []error
+}
+
+func NewScanner(path string) *Scanner {
+	var s Scanner
+
+	s.file = path
+	buf, err := ioutil.ReadFile(path)
+	if err != nil {
+		s.raise(err)
+		return &s
+	}
+
+	s.inBuf = bytes.NewBuffer(buf)
+
+	s.eof = false
+	s.Line = 1
+	s.Column = 1
+	s.offset = 0
+
+	return &s
+}
+
+const (
+	eofRune = -1
+	errRune = -2
+)
+
+func (s *Scanner) peek() rune {
+	if s.eof {
+		return eofRune
+	}
+
+	r, _, err := s.inBuf.ReadRune()
+	defer s.inBuf.UnreadRune() // rewind
```

Comment:
actually... we may not even need a `defer`. do we?
we're just interested in `r`, not the state of `s.inBuf` after we've read the rune.

Reply:
The point of peeking the scanner's input is to retrieve the next character as a look-ahead, not consume it. This allows a pretty neat combination with the `match` method.

If we don't unread the rune, this alters the state of the scanner's buffer, which defeats the idea of having a look-ahead rune.

Reply:
Totally agreeing on this.